cmd/snap-confine/ns-support.c

```diff
@@ -306,10 +307,149 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
+{
+	char base_snap_rev[PATH_MAX] = { 0 };
+	char fname[PATH_MAX] = { 0 };
+	char mnt_fname[PATH_MAX] = { 0 };
+	dev_t base_snap_dev;
+	int event_fd SC_CLEANUP(sc_cleanup_close) = -1;
+	pid_t parent, child;
+
+	// Read the revision of the base snap by looking at the current symlink.
+	sc_must_snprintf(fname, sizeof fname, "%s/%s/current",
+			 SNAP_MOUNT_DIR, base_snap_name);
+	if (readlink(fname, base_snap_rev, sizeof base_snap_rev) < 0) {
+		die("cannot read revision of base snap %s", fname);
+	}
+	if (base_snap_rev[sizeof base_snap_rev - 1] != '\0') {
+		die("cannot use symbolic link %s - value is too long", fname);
+	}
+	// Find the device that is backing the current revision of the base snap.
+	base_snap_dev = find_base_snap_device(base_snap_name, base_snap_rev);
+
+	// Store the PID of this process. This done instead of calls to getppid()
+	// below because then we can reliably track the PID of the parent even if
+	// the child process is re-parented.
+	parent = getpid();
+
+	// Create an eventfd for the communication with the child.
+	event_fd = eventfd(0, EFD_CLOEXEC);
+	if (event_fd < 0) {
+		die("cannot create eventfd for communication with inspection process");
+	}
+	// Fork a child, it will do the inspection for us.
+	child = fork();
+	if (child < 0) {
+		die("cannot fork support process for namespace inspection");
+	}
+
+	if (child == 0) {
+		// This is the child process which will inspect the mount namespace.
+		//
+		// Configure the child to die as soon as the parent dies. In an odd
+		// case where the parent is killed then we don't want to complete our
+		// task or wait for anything.
+		if (prctl(PR_SET_PDEATHSIG, SIGINT, 0, 0, 0) < 0) {
+			die("cannot set parent process death notification signal to SIGINT");
+		}
+		// Check that parent process is still alive. If this is the case then
+		// we can *almost* reliably rely on the PR_SET_PDEATHSIG signal to wake
+		// us up from eventfd_read() below. In the rare case that the PID
+		// numbers overflow and the now-dead parent PID is recycled we will
+		// still hang forever on the read from eventfd below.
+		debug("ensuring that parent process is still alive");
+		if (kill(parent, 0) < 0) {
+			switch (errno) {
+			case ESRCH:
+				debug("parent process has already terminated");
+				abort();
+			default:
+				die("cannot ensure that parent process is still alive");
+				break;
+			}
+		}
+		// Move to the mount namespace of the snap we're trying to inspect.
+		if (setns(mnt_fd, CLONE_NEWNS) < 0) {
+			die("cannot join the mount namespace in order to inspect it");
+		}
+		// Check if the namespace needs to be discarded.
+		bool should_discard = should_discard_current_ns(base_snap_dev);
+
+		// Send this back to the parent: 2 - discard, 1 - keep.
+		// Note that we cannot just use 0 and 1 because of the semantics of eventfd(2).
+		debug
+		    ("sending information about the state of the mount namespace");
+		if (eventfd_write(event_fd, should_discard ? 2 : 1) < 0) {
```

Comment:
I think I would prefer an enum here, so that it's easier to jump around using tags, eg:
```
typedef enum {
    SC_NS_INVALID = 0,
    SC_NS_KEEP,
    SC_NS_DISCARD, 
} sc_ns_discard_t
```

Reply:
I did something like that now.